mava/networks/sable_network.py

```diff
@@ -0,0 +1,547 @@
+
+        # Create dummy decay scale factor for FF Sable
+        if self.memory_config.type == "ff_sable":
+            self.memory_config.decay_scaling_factor = 1.0
+        assert (
+            self.memory_config.decay_scaling_factor >= 0
+            and self.memory_config.decay_scaling_factor <= 1
+        ), "Decay scaling factor should be between 0 and 1"
+
+        # Decay kappa for each head
+        self.decay_kappas = 1 - jnp.exp(
+            jnp.linspace(jnp.log(1 / 32), jnp.log(1 / 512), self.net_config.n_head)
+        )
+        self.decay_kappas = self.decay_kappas * self.memory_config.decay_scaling_factor
+
+        self.encoder = Encoder(
+            self.net_config,
+            self.memory_config,
+            self.n_agents_per_chunk,
+        )
+        self.decoder = Decoder(
+            self.net_config,
+            self.memory_config,
+            self.n_agents_per_chunk,
+            self.action_dim,
+            self.action_space_type,
+        )
+
+        # Set the executor and trainer functions
+        (
+            self.train_encoder_fn,
+            self.train_decoder_fn,
+            self.execute_encoder_fn,
+            self.autoregressive_act,
+        ) = self.setup_executor_trainer_fn()
+
+    def __call__(
+        self,
+        obs_carry: Observation,
+        action: chex.Array,
+        hstates: HiddenStates,
+        dones: chex.Array,
+        rng_key: Optional[chex.PRNGKey] = None,
+    ) -> Tuple[chex.Array, chex.Array, chex.Array]:
+        """Training phase."""
+        # Get the observation, legal actions, and timestep id
+        obs, legal_actions, timestep_id = (
+            obs_carry.agents_view,
+            obs_carry.action_mask,
+            obs_carry.step_count,
+        )
+        # Apply the encoder
+        v_loc, obs_rep, _ = self.train_encoder_fn(
+            encoder=self.encoder, obs=obs, hstate=hstates[0], dones=dones, timestep_id=timestep_id
+        )
+
+        # Apply the decoder
+        action_log, entropy = self.train_decoder_fn(
+            decoder=self.decoder,
+            obs_rep=obs_rep,
+            action=action,
+            legal_actions=legal_actions,
+            hstates=hstates[1],
+            dones=dones,
+            timestep_id=timestep_id,
+            rng_key=rng_key,
+        )
+
+        return v_loc, action_log, entropy
+
+    def get_actions(
+        self,
+        obs_carry: Observation,
+        hstates: HiddenStates,
+        key: chex.PRNGKey,
+    ) -> Tuple[chex.Array, chex.Array, chex.Array, HiddenStates]:
+        """Inference phase."""
+        # Get the observation, legal actions, and timestep id
+        obs, legal_actions, timestep_id = (
+            obs_carry.agents_view,
+            obs_carry.action_mask,
+            obs_carry.step_count,
+        )
+        # Decay the hidden states
+        decayed_hstates = jax.tree.map(
+            lambda x: x * self.decay_kappas[None, :, None, None, None], hstates
+        )
```

Comment:
Can we just store self.decay_kappas in the correct shape to start with, also `jnp.newaxis` please :pray:

Reply:
Nice idea